nav2_amcl/src/pf/pf.c

```diff
@@ -289,6 +293,77 @@ void pf_update_sensor(pf_t * pf, pf_sensor_model_fn_t sensor_fn, void * sensor_d
   }
 }
 
+double get_determinant(double *cov_matrix)
+{
+  double det = cov_matrix[0] * cov_matrix[4] * cov_matrix[8] + 
+              cov_matrix[1] * cov_matrix[5] * cov_matrix[6] + 
+              cov_matrix[3] * cov_matrix[7] * cov_matrix[2] -
+              cov_matrix[2] * cov_matrix[4] * cov_matrix[6] - 
+              cov_matrix[1] * cov_matrix[3] * cov_matrix[8] - 
+              cov_matrix[0] * cov_matrix[5] * cov_matrix[7];
+
+  return det;
+}
+
+int get_inverse(double *cov_matrix, double *result)
+{
+  double det = get_determinant(cov_matrix);
+  double inv[9] = {   1 * (cov_matrix[4] * cov_matrix[8] - cov_matrix[5] * cov_matrix[7]) / det,
+                        -1 * (cov_matrix[3] * cov_matrix[8] - cov_matrix[5] * cov_matrix[6]) / det,
+                         1 * (cov_matrix[3] * cov_matrix[7] - cov_matrix[4] * cov_matrix[6]) / det,
+                        -1 * (cov_matrix[1] * cov_matrix[8] - cov_matrix[2] * cov_matrix[7]) / det,
+                         1 * (cov_matrix[0] * cov_matrix[8] - cov_matrix[2] * cov_matrix[6]) / det,
+                        -1 * (cov_matrix[0] * cov_matrix[7] - cov_matrix[1] * cov_matrix[6]) / det,
+                         1 * (cov_matrix[1] * cov_matrix[5] - cov_matrix[2] * cov_matrix[4]) / det,
+                        -1 * (cov_matrix[0] * cov_matrix[5] - cov_matrix[2] * cov_matrix[3]) / det,
+                         1 * (cov_matrix[0] * cov_matrix[4] - cov_matrix[1] * cov_matrix[3]) / det};
+
+  memcpy(result, inv, 9*sizeof(double));
+
+  return 0;
+}
+
+int mult_1_3_x_3_3(double *mat_1_3, double *mat_3_3, double *result)
+{
+  double r[3] = { mat_1_3[0] * mat_3_3[0] + mat_1_3[1] * mat_3_3[3] + mat_1_3[2] * mat_3_3[6],
+                      mat_1_3[0] * mat_3_3[1] + mat_1_3[1] * mat_3_3[4] + mat_1_3[2] * mat_3_3[7],
+                      mat_1_3[0] * mat_3_3[2] + mat_1_3[1] * mat_3_3[5] + mat_1_3[2] * mat_3_3[8]};
+
+  memcpy(result, r, 3*sizeof(double));
+
+  return 0;
+}
+
+double mult_1_3_x_3_1(double *mat_1_3, double *mat_3_1)
+{
+  double result = mat_1_3[0] * mat_3_1[0] + mat_1_3[1] * mat_3_1[1] + mat_1_3[2] * mat_3_1[2];
+  return result;
+}
+
+// Return random std mean number
+double norm_random()
+{
+  double u = (double)rand()/RAND_MAX;
+  double v = (double)rand()/RAND_MAX;
+  double x = sqrt(-2*log(u))*cos(2*3.14159*v); // Check this
+
+  return x;
+
+}
+
+int generate_random_particle(double gps_x, double gps_y, double gps_yaw, double *cov_matrix, double *pose_v)
+{
+    double pose[3] = {norm_random(), norm_random(), norm_random()};
+  
+    mult_1_3_x_3_3(pose, cov_matrix, pose);
+    pose[0] = pose[0] + gps_x;
+    pose[1] = pose[1] + gps_y;
+    pose[2] = pose[2] + gps_yaw;
+
+    memcpy(pose_v, pose, 3*sizeof(double));
```

Comment:
just assign directly into `pose_v`?  also do you want to remove the `gps_` prefix?

Reply:
`gps` prefix is redundant, right

+ see the comment above on `memcpy`